platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,173 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.state.service;
+
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.state.primitives.ProtoBytes;
+import com.hedera.hapi.node.state.roster.Roster;
+import com.hedera.hapi.node.state.roster.RosterState;
+import com.hedera.hapi.node.state.roster.RosterState.Builder;
+import com.hedera.hapi.node.state.roster.RoundRosterPair;
+import com.hedera.hapi.platform.state.PlatformState;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.common.RosterStateId;
+import com.swirlds.platform.roster.RosterUtils;
+import com.swirlds.platform.roster.RosterValidator;
+import com.swirlds.state.spi.CommittableWritableStates;
+import com.swirlds.state.spi.WritableKVState;
+import com.swirlds.state.spi.WritableSingletonState;
+import com.swirlds.state.spi.WritableStates;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Provides write methods for interacting with Rosters.
+ */
+public class WritableRosterStore extends ReadableRosterStoreImpl {
+
+    private final WritableStates writableStates;
+
+    /**
+     * The roster state singleton. This is the state that holds the candidate roster hash and the list of pairs of
+     * active roster hashes and the round number in which those rosters became active.
+     *
+     * @implNote the use of {@link ReadablePlatformStateStore} and {@link WritablePlatformStateStore} to provide access
+     * to the roster states (beyond just the {@link PlatformState}) is deliberate, for convenience.
+     */
+    private final WritableSingletonState<RosterState> rosterState;
+
+    private final WritableKVState<ProtoBytes, Roster> rosterMap;
+
+    /**
+     * Constructs a new {@link WritableRosterStore} instance.
+     *
+     * @param writableStates the readable states
+     */
+    public WritableRosterStore(@NonNull final WritableStates writableStates) {
+        super(writableStates);
+        this.writableStates = writableStates;
+        this.rosterState = writableStates.getSingleton(RosterStateId.ROSTER_STATES_KEY);
+        this.rosterMap = writableStates.get(RosterStateId.ROSTER_KEY);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setActiveRoster(@NonNull final Roster roster, final long round) {
+        Objects.requireNonNull(roster);
+        RosterValidator.validate(roster);
+
+        final RosterState previousRosterState = rosterStateOrThrow();
+        Objects.requireNonNull(previousRosterState);
+
+        // update the roster state
+        final List<RoundRosterPair> roundRosterPairs = new LinkedList<>(previousRosterState.roundRosterPairs());
+        final Bytes activeRosterHash = RosterUtils.hashOf(roster).getBytes();
+        roundRosterPairs.addFirst(new RoundRosterPair(round, activeRosterHash));
```

Comment:
As I [stated](https://github.com/hashgraph/hedera-services/pull/15362#discussion_r1747248541) in my previous review, this implementation will cause this list to grow indefinitely.

According to the original discussions, and as written in the design proposal, this should not be the case. You need to keep track of the previous address book, but not the previous previous address book, etc..

Reply:
Your comment in that review asked to overwrite the roster, which seems incorrect to me (see my reply).

One option I considered is to remove the last element in the list each time one is added. I however didn't add that because:
1. Removal of the last pair (near term) wasn't described in the proposal.
2. In the near term (DAB 2.5?) which this change is for, adoption is only at upgrade boundaries so this code will be called only once per upgrade.
3. Long term, full DAB, adoption is once a day. Before we reach here however, more work will be required in this method e.g. swapping the List for a better data structure.

That said, I'm open to removing the last pair now if we agree. @poulok @edward-swirldslabs @anthony-swirldslabs thoughts?

Reply:
At this stage of implementation, this list should only store 2 roster round pairs: The active roster and the previously active roster. Even with a new roster being adopted only on upgrades, the list should not be allowed to grow unbounded over time. If we wanted to be super safe, we should add code that removes roster round pairs from the end of the list that are for ancient rounds, which is what the DAB 3 implementation will need to be.